### PR TITLE
[DISCO-2791] Handle requests with weather location key

### PR DESCRIPTION
**Pull request by tiftran**
## References

JIRA: [DISCO-2791](https://mozilla-hub.atlassian.net/browse/DISCO-2791)

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [ ] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [ ] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable)
- [ ] `[load test: (abort|warn)]` keywords are applied (if applicable)
- [ ] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [ ] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)

[DISCO-2791]: https://mozilla-hub.atlassian.net/browse/DISCO-2791?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by tiftran** [comment]
Ripped out the extra get_location_via_location key, just wanted to check in to see how it looks to everyone
I think I still have tests to write relating to the cache hits.

File: merino/providers/weather/backends/accuweather.py [hunk 1]
```diff
@@ -894,6 +957,51 @@ async def get_weather_report(self, geolocation: Location) -> WeatherReport | Non
         Raises:
             AccuweatherError: Failed request or 4xx and 5xx response from AccuWeather.
         """
+        # Look up for all the weather data from the cache.
+        try:
+            with self.metrics_client.timeit("accuweather.cache.fetch"):
```
Comment (ncloudioj) [inline merino/providers/weather/backends/accuweather.py:962]:
```suggestion
            with self.metrics_client.timeit("accuweather.cache.fetch-via-location-key"):
```

File: merino/providers/weather/backends/accuweather.py [hunk 2]
```diff
@@ -894,6 +957,51 @@ async def get_weather_report(self, geolocation: Location) -> WeatherReport | Non
...
+                cached_data: list[bytes | None] = await self.cache.run_script(
+                    sid=SCRIPT_LOCATION_KEY_ID,
+                    keys=[location_key],
+                    # The order matters below. See `LUA_SCRIPT_CACHE_BULK_FETCH` for details.
```
Comment (ncloudioj) [inline merino/providers/weather/backends/accuweather.py:966]:
```suggestion
                    # The order matters below. See `LUA_SCRIPT_CACHE_BULK_FETCH_VIA_LOCATION` for details.
```

File: tests/unit/providers/weather/backends/test_accuweather_city.py [hunk 3]
```diff
@@ -188,6 +188,12 @@ def fixture_accuweather_city_parameters(
     }
 
 
+@pytest.fixture(name="accuweather_location_key")
+def fixture_accuweather_location_key() -> str:
+    """Location key for the expected weather report."""
+    return str(39376)
```
Comment (ncloudioj) [inline tests/unit/providers/weather/backends/test_accuweather_city.py:194]:
```suggestion
    return "39376"
```

**Review by ncloudioj — COMMENTED** [review]
Looks great, just a few minor comments.

File: merino/providers/weather/backends/accuweather.py [hunk 4]
```diff
@@ -894,6 +957,51 @@ async def get_weather_report(self, geolocation: Location) -> WeatherReport | Non
...
+                    args=[
+                        self.cache_key_for_accuweather_request(
+                            self.url_current_conditions_path.format(
+                                location_key=location_key
+                            )
+                        ),
+                        self.cache_key_for_accuweather_request(
+                            self.url_forecasts_path.format(location_key=location_key)
+                        ),
+                        self.url_location_key_placeholder,
```
Comment (ncloudioj) [inline merino/providers/weather/backends/accuweather.py:976]: This placeholder can be dropped as well.

File: tests/unit/providers/weather/backends/test_accuweather_city.py (hunk 3 above)
Comment (tiftran) [reply tests/unit/providers/weather/backends/test_accuweather_city.py]: yikes 🙈

File: merino/providers/weather/backends/accuweather.py [hunk 5]
```diff
@@ -793,7 +837,9 @@ async def store_request_into_cache(
 
         return cache_ttl.seconds
 
-    def emit_cache_fetch_metrics(self, cached_data: list[bytes | None]) -> None:
+    def emit_cache_fetch_metrics(
+        self, cached_data: list[bytes | None], location_key=None
```
Comment (ncloudioj) [inline merino/providers/weather/backends/accuweather.py:841]: How about replacing `location_key` with `skip_location_key` as a boolean? Whichever you prefer, let's add this to the `Params` section of the docstring.

**Review by ncloudioj — APPROVED** [review]
LGTM, r=wc.

**Review by Herraj — APPROVED** [review]
🚀

**Pull request merged**